backend/server/db/helpers/models.py

```diff
@@ -0,0 +1,115 @@
+from typing import Dict, List, Literal, NewType, Optional, Union
+from uuid import UUID
+from pydantic import BaseModel
+
+#
+# NewTypes
+#
+SessionID = NewType('SessionID', UUID)
+SessionToken = NewType('SessionToken', str)
+RefreshToken = NewType('RefreshToken', str)
+
+#
+# User Storage Models
+#
+class UserDegreeStorage(BaseModel):
+    programCode: str
+    specs: List[str]
+
+
+class UserCourseStorage(BaseModel):
+    code: str
+    suppressed: bool
+    mark: Union[Literal['SY', 'FL', 'PS', 'CR', 'DN', 'HD'], int, None]
+    uoc: int
+    ignoreFromProgression: bool
+
+# TODO-OLLI(pm): https://docs.pydantic.dev/latest/concepts/models/#rootmodel-and-custom-root-types
+# we can make this a RootModel or use a TypeAdapter if we want stronger type checks here
+type UserCoursesStorage = Dict[str, UserCourseStorage]
+
+class YearTerm(BaseModel):
+    Y: int
+    T: int
+
+class PlannerYear(BaseModel):
+    T0: List[str]
+    T1: List[str]
+    T2: List[str]
+    T3: List[str]
+
+class UserPlannerStorage(BaseModel):
+    unplanned: List[str]
+    startYear: int
+    isSummerEnabled: bool
+    mostRecentPastTerm: YearTerm
+    years: List[PlannerYear]
+    lockedTerms: Dict[str, bool]
+
+class _BaseUserStorage(BaseModel):
+    # NOTE: could also put uid here if we want
+    guest: bool
+
+class UserStorage(_BaseUserStorage):
+    setup: Literal[True] = True
+    degree: UserDegreeStorage
+    courses: UserCoursesStorage
+    planner: UserPlannerStorage
+
+class NotSetupUserStorage(_BaseUserStorage):
+    setup: Literal[False] = False
+
+# https://github.com/pydantic/pydantic/issues/1223#issuecomment-1152323275
+class PartialUserStorage(BaseModel):
+    # for batch user storage update
+    # TODO-OLLI(pm): remove this and use keyword args since they cannot be None
```

Comment:
Is this comment
1. The fields should no longer be none?
2. The arguments passed in should not be allowed to be `None` but, the argument can be omitted?

Reply:
Im still on fence with this.
The aim was to allow just passing in Degree or Courses or a combo of any... But with Pydantic, to indicate a not required field, you must also allow that field to be `None`. Because of this, In Pydantics deser, something like `PartialUserStorage(courses=...)` would convert to a dict with only a `courses` key, BUT, something like `PartialUserStorage(courses=..., degree=None)` converts to a dict including the `degree` field (BAD). So with this Pydantic way, I have to filter out the None fields after deserializing...

So this comment is more just saying "I don't actually want there to be `None` fields". 

With keyword args, you can do something like `def func(*, a: int, b: int): ...` which truly allows you to have optional fields or params that are not allowed to be given as `None`, thus removing any unclear behaviour.

Reply:
but tbh, i think the post deserialization filter is just cleanest, and gives the type validation i want, whereas keyword args would not (specifically for the courses dict), without using the pydantic validation decorators (which are more added code complexity)